core/src/main/scala/akka/persistence/r2dbc/internal/postgres/PostgresDurableStateDao.scala

```diff
@@ -506,13 +507,18 @@ private[r2dbc] class PostgresDurableStateDao(executorProvider: R2dbcExecutorProv
       }
     }
 
-    result.map { case (updatedRows, changeEventTimestamp) =>
-      if (updatedRows != 1)
-        throw new IllegalStateException(
-          s"Update failed: durable state for persistence id [${state.persistenceId}] could not be updated to revision [${state.revision}]")
-      else {
+    result.flatMap { case (updatedRows, changeEventTimestamp) =>
+      if (updatedRows != 1) {
+        if (settings.migration) {
+          // during migration,
+          upsertState(state, value, changeEvent, insertRevisionOverride = true)
```

Comment:
I think it would be more clean to have a separate dao method that is used from migration. It can do a real upsert instead of the insert/update attempts. It would not have to deal with changeEvent and it should not use the assert-single-writer check.

Reply:
A real upsert does not work either, since the `ON CONFLICT(persistence_id,revision)` only works on exact matches to decide if to insert or to update.

In our case, want to update if there is row with the same persistence id and a an `existing revision = new revision-1`

Reply:
why do we need to include the revision in the upsert condition? can't we use `ON CONFLICT(persistence_id)`?

> existing revision = new revision-1

I don't think that is true for migration, and the reason why the instructions were to disable the single-writer assert.

Reply:
For a `ON CONFLICT(persistence_id)` definition, I got:

```
io.r2dbc.postgresql.ExceptionFactory$PostgresqlBadGrammarException, with message: there is no unique or exclusion constraint matching the ON CONFLICT specification. (MigrationToolSpec.scala:416)
```

Reply:
But deciding on a insert / update should be doable with a simple upfront query, with minimal refactoring of the existing code in `PostgresDurableStateDao`.